### PR TITLE
Added EPA data

### PR DESCRIPTION
**Pull request by cosmostevens**
- Added EPA Data
- @lindaicing  Added JavaScript variable "greenPowerPercentage"
- @amtims Adjusted the calculation feature so that the percentage of electricity a company gets from green energy sources is taken into account. The algorithm is:
If the percentage of green energy is not 0
     Then there environmental impact score is 4 + (green energy percentage X 2.5)
If this results in a number greater than 9
      The environmental impact is set to 9

Questions:
1. Do these numbers work? Some of these companies have percentages above 100% Im assuming because they provide green energy to other companies. This would mean you would need to have 200% to get a perfect 9. Since there are other environmental factors that we do not yet have data on, I did not want to make a 9 score too easy. On the other side of things, some of these companies have a 10% green percentage. In Sacramento, SMUD (the electricity provider) gives customers an option to have a minimum of 50% percent of their electricity come from green energy alternatives for a flat rate cost of about $2 a month. So hats off to the companies who have started to transition to greener alternatives, but if I have a better green energy percentage for only $2 a month than they do, then I am not very impressed by their efforts, therefore I feel they should still get a reasonably low score.
2. How would you like this percentage to play into overall score? I did not add anything in yet, but I can add a point here or there. If you are not sure I can come up with something and put it into a new request.

### CODE REVIEW

**Comment by amtims** [comment]
In response to Question 1: If it's really easy to use 50% green energy maybe it should be something like "If the percentage of green energy is greater than 50%" because of how low the bar is. From that point, even if they only use 50% they'll still only get a 4.5 which isn't great. The other point is that companies generate much more energy for a household, so I think it could be a lot more expensive to buy renewable energy. We could look into how much these companies are paying for different energy percentages. 

Q2: If a company has multiple variables, maybe they should be averaged for now for it's final score. If IKEA gets a Bcorp score hypothetically of 7.0, but the EPA score is 9, then we could give it an 8.0. Thoughts?

**Comment by cosmostevens** [comment]
- Added More categories from the BCorp Data: Community, Environment, Governance, & Workers.
- Reworked Calculation feature as follows:
*.    IF there is no manual Score
*.         IF there is a greenPowerPercentage we give it a rating from 4 to 10 
*.         IF there is a bcorpEnvironmentScore we give it a rating out of 10 (the mean environment score will give them roughly a 7.5)
*.         IF they are a Bluesign Partner we give a rating of 10 (Bluesign partnership tells us that their materials are sustainable)
*.         We then take an average of the factors which we have data for. (if greenpower is 6, there is no bcorp score, and they are a blusign partner they get a (6 + 10) / 2 = 8
*.         IF their score has one factor, the highest thy can get is a 7.
*.         IF their score has two factors, the highest thy can get is a 8.
*.         IF their score has three+ factors, the highest thy can get is a 9.
(This way they can get a stronger score if we have more data sources backing up good behavior)
(The same strategy is repeated for labor but we only have one labor source thus far, and the same strategy will be repeated for Animal welfare when we have data for that category)
*.         We then take an average of nonzero category data which becomes the overall data
*.         IF we only have one category of data (eg. just labor) the highest overall score they can get is a 7
*.         IF they have two categories of data, they highest they can get is an 8.
*.         IF they have all three, the highest they can get is a 9.

Note: 
I wrote a code to find important data on the different bcorp categories so that we could find out what their data actually means. The results I got will explain the different numbers chosen for the bcorp related calculations. The goal was to give 7 - 7.5 ratings to companies whose bcorp scores were at their respective averages.

BCorp Stats:
Community: 
	Min:	4.4
	Max: 125.4
	Avg:	30.547619402985102
Environment: 
	Min:	0.1
	Max: 96.9
	Avg:	17.79896268656722
Governance: 
	Min:	1.9
	Max: 41.3
	Avg:	12.51308955223879
Workers: 
	Min:	0.7
	Max: 81.8
	Avg:	20.796514925373167
Overall: 
	Min:	78.2
	Max: 184.0
	Avg:	96.5093880597013

**Comment by amtims** [comment]
Hey Cosmo, Billy and I talked and I think we are thinking of:

- removing the 7/8/9 caps
- not scoring based on number of certifications (since some companies can't afford to and might not be a good indicator)
- capping at 9.5 instead of 9 for all scores.

Could you make these changes? If you have any questions I'm free the rest of the day if you want to work though it again over a meeting if need be.
Thank you!

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 1]
```diff
@@ -342,25 +409,74 @@ public void update(Business business){
...
+            if (environmentalImpactScore != 0 && environmentalImpactFactors != 0) {
+                this.environmentScore = environmentalImpactScore / environmentalImpactFactors;
+            }
+            if (this.environmentScore > 9.5) {
+                this.environmentScore = 9.5;
+            }
+            // Labor Practice Score
+            double laborImpactScore = 0;
+            int laborImpactFactors = 0;
+
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:436]: Remove newline here

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 2]
```diff
@@ -342,25 +409,74 @@ public void update(Business business){
...
+            // Environmental Impact Score
+            double environmentalImpactScore = 0;
+            int environmentalImpactFactors = 0;
+            if (this.greenPowerPercentage != 0) {
+                ++environmentalImpactFactors;
+                environmentalImpactScore += (4 + ((this.greenPowerPercentage / 100) * 2.5));
+            }
+            if (this.bcorpEnvironmentScore != 0) {
+                ++environmentalImpactFactors;
+                environmentalImpactScore += (this.bcorpEnvironmentScore / 2.4);
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:421]: Where did 2.4 come from?

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 3]
```diff
@@ -342,25 +409,74 @@ public void update(Business business){
...
     public void calculate() {
         if (this.overallScore == 0) {
-            if (this.bcorpScore != 0) {
-                this.overallScore = (this.bcorpScore / 13);
+            // Environmental Impact Score
+            double environmentalImpactScore = 0;
+            int environmentalImpactFactors = 0;
+            if (this.greenPowerPercentage != 0) {
+                ++environmentalImpactFactors;
+                environmentalImpactScore += (4 + ((this.greenPowerPercentage / 100) * 2.5));
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:417]: Where did the 4 come from? Where did 2.5 come from?

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 4]
```diff
@@ -342,25 +409,74 @@ public void update(Business business){
...
+            if (this.environmentScore > 9.5) {
+                this.environmentScore = 9.5;
+            }
+            // Labor Practice Score
+            double laborImpactScore = 0;
+            int laborImpactFactors = 0;
+
+            if (this.bcorpWorkerScore != 0) {
+                ++laborImpactFactors;
+                laborImpactScore += (this.bcorpWorkerScore / 2.7);
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:439]: Where did 2.7 come from?

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 5]
```diff
@@ -342,25 +409,74 @@ public void update(Business business){
...
+                ++environmentalImpactFactors;
+                environmentalImpactScore += (4 + ((this.greenPowerPercentage / 100) * 2.5));
+            }
+            if (this.bcorpEnvironmentScore != 0) {
+                ++environmentalImpactFactors;
+                environmentalImpactScore += (this.bcorpEnvironmentScore / 2.4);
+            }
+            if (bluesignPartner) {
+                ++environmentalImpactFactors;
+                environmentalImpactScore += 10;
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:425]: Bluesign partners get a 10/10 for environment!? If bluesign is that prestigious then great, but a 10/10 is high enough to force me to require a double check on this.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 1 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:436]: what is a double?

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 2 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:421]: 2.4 weights the data properly because I think they are out of 240, so dividing by 2.4 allows it to be on our scale of 10

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 1 above)
Comment (CubeOfCheese) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:436]: 5 is an int or integer
5.4 is a double or float

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 3 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:417]: 4/10 is the floor score of what a certification-holding company can get, and 2.5 is the highest boost (a 6.5 rating) a company can get for using 100% green energy. If they produce for than 100% (like Ikea) they get more points. Ikea would get probably a 9.5 for this

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 4 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:439]: @cosmostevens is this supposed to be 2.4?

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 2 above)
Comment (CubeOfCheese) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:421]: Excellent thank you. This needs to be added as inline documentation. Something as simple as // Bcorp subscores are out of 240

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 5 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:425]: I'm not sure what this is. @cosmostevens Maybe we discussed this yesterday but I can't remember what this part was about

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 1 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:436]: Oh okay. I wanted the sub categories to be doubles for more accurate calculating. Does this answer the question? I guess they are for mathematical conversion rather than a weight.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 3 above)
Comment (amtims) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:417]: @cosmostevens maybe the 4 should be a 5 instead... if they use 100% green energy a 6.5 might actually be low for their environmental score. That way using 100% green energy could get them 7.5.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 2 above)
Comment (cosmostevens) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:421]: Sorry i should have put a comment in. I will also make a variable explaining what the 2.4 stands for. The BCorp scores aren't out of 240. I wrote a program to find the statistical data associated with the different categories (avg, nonzero min, and max). I put the results in a previous comment if you want to look at it. The goal for scoring was to give a 7 - 7.5 environmentalImpactScore for a company who got a score around the mean of that specific category(int this case the mean is 17.8). To do that for the bcorpEnvironmentalScore category you have to divide it by 2.4. That way if they get a 24+ they will get a 10 in this category, a 17.8 will get a 7.4, a 12 will get a 5, and so on.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 3 above)
Comment (cosmostevens) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:417]: Sounds good! I will switch that up.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 5 above)
Comment (cosmostevens) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:425]: I wrote it this way so that it could only help their score. It worked better, before we took out the portion of code that capped ratings depending on how many sources we have on them. With the caps it meant that if a company just had a bluesign badge they would get a 7, if they had a bluesign badge and another factor they could get an 8, and so on. However without the caps, you're totally right, this doesn't make sense anymore. I can switch it to something else. We will discuss this in the meeting today, I have a couple ideas about where to go with this.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java (hunk 2 above)
Comment (CubeOfCheese) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:421]: I am very skeptical of this. Was it decided that the mean score should result in a 7-7.5? This calls into question if the mean should even be a factor.

**Comment by CubeOfCheese** [comment]
Okay Also! This PR needs to update the extension as well because some of these calculations are happening in the extension still.

File: Extension/background.js [hunk 6]
```diff
@@ -38,23 +38,14 @@ function reloadExt(request, sender, sendResponse) {
                     }
                 }
 
-                if(notBlacklisted){
```
Comment (lindaicing) [inline Extension/background.js:41]: This is just calculating for the backend right? Since the calculation is already calculated before the JSON is parsed, this is just removing the js calculations?

Comment (CubeOfCheese) [reply Extension/background.js:41]: Yes that was the intention. It looks like I accidentally changed the url for the api call. I'll fix that right now

Comment (CubeOfCheese) [reply Extension/background.js]: Okay I think I fixed

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java [hunk 7]
```diff
@@ -342,25 +409,81 @@ public void update(Business business){
 
     public void calculate() {
         if (this.overallScore == 0) {
-            if (this.bcorpScore != 0) {
-                this.overallScore = (this.bcorpScore / 13);
+            // Environmental Impact Score
+            double environmentalImpactScore = 0;
+            double bcorpEnvironmentMean = 17.8; // Mean of all bcorpWorkerScore data
```
[the remark this answers is not in the document]
Comment (CubeOfCheese) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:414]: @cosmostevens

[the remark this answers is not in the document]
Comment (cosmostevens) [reply Website-and-Backend/src/main/java/com/example/appengine/springboot/Business.java:414]: Oh snap, just saw this.

File: Website-and-Backend/src/main/java/com/example/appengine/springboot/SpringbootApplication.java [hunk 8]
```diff
@@ -37,7 +38,8 @@ public static void main(String[] args) {
 
     // Controls Searches of all Data Sources
     @GetMapping("/score/{company}")
-    public Business masterController(@PathVariable("company") String companyName) throws IOException {
+    public Business masterController(@PathVariable("company") String companyName, HttpServletResponse response) throws IOException {
+        response.addHeader("Access-Control-Allow-Origin", "*");
```
Comment (CubeOfCheese) [inline Website-and-Backend/src/main/java/com/example/appengine/springboot/SpringbootApplication.java:42]: This change is not related to this PR at all. This change can be in its own PR. Please remove
